examples/pytorch/graphsage/multi_gpu_node_classification.py

```diff
@@ -31,34 +48,37 @@ def forward(self, blocks, x):
                 h = self.dropout(h)
         return h
 
-    def inference(self, g, device, batch_size, num_workers, buffer_device=None):
-        # The difference between this inference function and the one in the official
-        # example is that the intermediate results can also benefit from prefetching.
-        g.ndata['h'] = g.ndata['feat']
-        sampler = dgl.dataloading.MultiLayerFullNeighborSampler(1, prefetch_node_feats=['h'])
-        dataloader = dgl.dataloading.DataLoader(
-                g, torch.arange(g.num_nodes()).to(g.device), sampler, device=device,
-                batch_size=1000, shuffle=False, drop_last=False, num_workers=num_workers,
-                persistent_workers=(num_workers > 0))
-        if buffer_device is None:
-            buffer_device = device
+    def inference(self, g, device, batch_size,
+                  buffer_device, q):
```

Comment:
I feel the example may need more explanations. How about adding some descriptions in the example docstring to cover: (1) how training is distributed? and (2) how inference is conducted?

Reply:
Besides, I don't like having this `q` in user code.  Since `q` is solely for sharing tensors via shared memory, is it possible to organize the code so the users don't have to be aware of a queue?  Other ways of sharing tensors could also work.

Reply:
Not sure if a better solution exist, but my solution explicitly creates a shared memory object
https://github.com/dmlc/dgl/blob/6752bd45155139673fa4b3c2f199f6e893660f8e/python/dgl/dataloading/dataloader.py#L206-L225
Probably we need to wrap it into a utility function.

Reply:
Another solution could be wrapping `torch.multiprocessing.spawn` function as `dgl.multiprocessing.spawn` which creates a queue in the master process and the worker processes stores the queue object into a global variable.  Then we have some function like `call_once_and_share(func)` where only one worker executes the function `func` and distributes the result to other workers.

Reply:
@BarclayII I switched it to using `dgl.utils.*_shared_mem_array()` and removed the queue.

@jermainewang I've added a docstring to the `inference()` and `shared_tensor()` functions, as well as a block of comments to describe how work is distributed for training.

Reply:
@BarclayII Actually upon further testing, I'm seeing errors when the mmap'd memory gets freed:
```
terminate called after throwing an instance of 'dmlc::Error'
  what():  [01:44:07] /workspace/dgl/src/runtime/shared_mem.cc:58: Check failed: munmap(ptr_, size_) != -1: Invalid argument
Stack trace:
  [bt] (0) /usr/local/lib/python3.8/dist-packages/dgl-0.9-py3.8-linux-x86_64.egg/dgl/libdgl.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x80) [0x7fa5f814ad70]
  [bt] (1) /usr/local/lib/python3.8/dist-packages/dgl-0.9-py3.8-linux-x86_64.egg/dgl/libdgl.so(dgl::runtime::SharedMemory::~SharedMemory()+0x105) [0x7fa5f851cfa5]
  [bt] (2) /usr/local/lib/python3.8/dist-packages/dgl-0.9-py3.8-linux-x86_64.egg/dgl/libdgl.so(std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()+0x48) [0x7fa5f81246f8]
  [bt] (3) /usr/local/lib/python3.8/dist-packages/dgl-0.9-py3.8-linux-x86_64.egg/dgl/libdgl.so(+0x27486d) [0x7fa5f806586d]
  [bt] (4) /usr/local/lib/python3.8/dist-packages/dgl-0.9-py3.8-linux-x86_64.egg/dgl/libdgl.so(DGLArrayAllocSharedMem+0x16c) [0x7fa5f85161cc]
  [bt] (5) /usr/lib/x86_64-linux-gnu/libffi.so.7(+0x6ff5) [0x7fa77c6a3ff5]
  [bt] (6) /usr/lib/x86_64-linux-gnu/libffi.so.7(+0x640a) [0x7fa77c6a340a]
  [bt] (7) /usr/lib/python3.8/lib-dynload/_ctypes.cpython-38-x86_64-linux-gnu.so(_ctypes_callproc+0x5b6) [0x7fa77c6bc306]
  [bt] (8) /usr/lib/python3.8/lib-dynload/_ctypes.cpython-38-x86_64-linux-gnu.so(+0x139dc) [0x7fa77c6bc9dc]


Traceback (most recent call last):
  File "multi_gpu_node_classification.py", line 221, in <module>
    mp.spawn(train, args=(n_procs, graph, num_classes, split_idx), nprocs=n_procs)
  File "/usr/local/lib/python3.8/dist-packages/torch/multiprocessing/spawn.py", line 230, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/usr/local/lib/python3.8/dist-packages/torch/multiprocessing/spawn.py", line 188, in start_processes
    while not context.join():
  File "/usr/local/lib/python3.8/dist-packages/torch/multiprocessing/spawn.py", line 130, in join
    raise ProcessExitedException(
torch.multiprocessing.spawn.ProcessExitedException: process 7 terminated with signal SIGABRT
/usr/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
``` 
Any ideas?

Reply:
How did you throw this error?